x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java

```diff
@@ -0,0 +1,282 @@
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.RATE_LIMITED;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.REPLAY_NONCE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.RETRY_AFTER;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.TYPE;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.USER_ACTION_REQUIRED;
+import static org.wildfly.security.certificate.management.x500.cert.acme.Acme.USER_AGENT;
+import static org.wildfly.security.certificate.management.x500.cert.acme.CertMgmtMessages.acme;
+
+public abstract class AsyncAcmeClientSpi {
+
+    private static final long DEFAULT_RETRY_AFTER_MILLI = 3000;
+    private static final String USER_AGENT_STRING = "Elytron ACME Client/" + "1.0.0.Alpha1-SNAPSHOT";
+    private final AcmeClientSpi delegate;
+
+    public AsyncAcmeClientSpi() {
+        AsyncAcmeClientSpi impl = this;
+
+        this.delegate = new AcmeClientSpi() {
+            @Override
+            public AcmeChallenge proveIdentifierControl(AcmeAccount account, List<AcmeChallenge> challenges) throws AcmeException {
+                return impl.proveIdentifierControl(account, challenges);
+            }
+
+            @Override
+            public void cleanupAfterChallenge(AcmeAccount account, AcmeChallenge challenge) throws AcmeException {
+                impl.cleanupAfterChallenge(account, challenge);
+            }
+        };
+    }
+
+    protected abstract AcmeChallenge proveIdentifierControl(AcmeAccount account, List<AcmeChallenge> challenges) throws AcmeException;
+
+    protected abstract void cleanupAfterChallenge(AcmeAccount account, AcmeChallenge challenge) throws AcmeException;
+
+    public Uni<byte[]> getNewNonce(final AcmeAccount account, final boolean staging) {
+        if (account == null) {
+            return Uni.createFrom().failure(new IllegalArgumentException("account"));
+        }
+        HttpClient httpClient = HttpClient.newHttpClient();
+
+        return getResourceUrl(account, AcmeResource.NEW_NONCE, staging)
+                .chain((URL newNonceUrl) -> {
+                    final URI newNonceUri;
+                    try {
+                        newNonceUri = newNonceUrl.toURI();
+                    } catch (URISyntaxException e) {
+                        throw new RuntimeException(e);
+                    }
+                    HttpRequest httpRequest = HttpRequest.newBuilder().method("HEAD", HttpRequest.BodyPublishers.noBody())
+                            .uri(newNonceUri)
+                            .header(ACCEPT_LANGUAGE, Locale.getDefault().toLanguageTag())
+                            .header(USER_AGENT, USER_AGENT_STRING).build();
+                    return Uni.createFrom().future(httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString()));
```

Comment:
That's exactly why we do not want to use this HTTP client. 

sendAsync is not emitting on an uncontrolled F/J thread. 
Also, creating a Uni from a future may be blocking (if not an instance of CompletionStage, which in this case is not a problem).

Reply:
`Future` are _blocking_, in the sense you cannot listen for the outcome without blocking. So, not great.

`CompletionStage` is the best you can do (or returns a `Uni`). At least you can listen for the result without blocking and we can create instances of Uni from `CompletionStage`. 

However, makes sure you never use the `async` method from the CompletionStage API.

Reply:
@Skyllarr @cescoffier Should `Uni` be indeed introduced right now in the client SPI, since `Uni` is already featuring in other parts of this project ?

Reply:
@cescoffier Sounds good, I have updated the `HttpClientSpi` to return the `CompletableFuture` instead of the `Future`, since `CompletableFuture` implements `CompletionStage`. Thanks for explaining!

@sberyozkin `Uni` instances are being returned in the `AsyncAcmeClientSpi`. This is so that this SPI is not blocking and we can handle the failures with `onFailure()` there for example. Which other parts of the project you had in mind that would introduce `Uni` instead of the `AsyncAcmeClientSpi`? It is this SPI that should be used in Quarkus code